internal/pkg/otel/manager/execution_subprocess.go

```diff
@@ -212,6 +221,35 @@ func (r *subprocessExecution) startCollector(ctx context.Context, logger *logger
 	return ctl, nil
 }
 
+func (r *subprocessExecution) getCollectorPorts() (healthCheckPort int, metricsPort int, err error) {
+	randomPorts := make([]*int, 0, 2)
+	// if the ports are defined (non-zero), use them
+	if r.collectorMetricsPort == 0 {
+		randomPorts = append(randomPorts, &metricsPort)
+	} else {
+		metricsPort = r.collectorMetricsPort
+	}
+	if r.collectorHealthCheckPort == 0 {
+		randomPorts = append(randomPorts, &healthCheckPort)
+	} else {
+		healthCheckPort = r.collectorHealthCheckPort
+	}
+
+	if len(randomPorts) == 0 {
+		return healthCheckPort, metricsPort, nil
+	}
+
+	// we need at least one random port, create it
+	ports, err := findRandomTCPPorts(len(randomPorts))
+	if err != nil {
+		return 0, 0, err
+	}
+	for i, port := range ports {
+		*randomPorts[i] = port
+	}
+	return healthCheckPort, metricsPort, nil
+}
```

Comment:
Nit: would this be a simpler / easier to follow implementation?


```suggestion
func (r *subprocessExecution) getCollectorPorts() (healthCheckPort int, metricsPort int, err error) {
	if r.collectorHealthCheckPort > 0 {
		healthCheckPort = r.collectorHealthCheckPort
	} else {
		ports, findPortsErr := findRandomTCPPorts(1)
		if findPortsErr != nil {
			return 0, 0, err
		}
		healthCheckPort = ports[0]
	}

	if r.collectorMetricsPort > 0 {
		metricsPort = r.collectorMetricsPort
	} else {
		ports, findPortsErr := findRandomTCPPorts(1)
		if findPortsErr != nil {
			return 0, 0, err
		}
		metricsPort = ports[0]
	}

	return healthCheckPort, metricsPort, nil
}
```

It does require calling `findRandomTCPPorts` twice.  Maybe you were trying to avoid that for some reason?

Reply:
Yes, the reason is that we can be sure not to have a conflict between the two. The logic unfortunately has four different cases that all need to fit under one `findRandomTCPPorts` call. I'll be very happy to change it if you can show me a clearer implementation.

Reply:
Yes calling `findRandomTCPPorts` twice with it's current implementation introduces the low but not impossible risk it allocates the same port twice. 

In the current implementation, it holds the listener bound to each successively allocated port until it has allocated the requested amount which prevents this collision from happening